lib/charms/blackbox_k8s/v0/blackbox_probes.py

```diff
@@ -0,0 +1,634 @@
+        A probe in blackbox is defined by a module and a static_config target. Those
+        are then organised in a prometheus job for proper scraping.
+        The `BlackboxProbesProvider` constructor requires
+        the name of the relation over which a probe target
+        is exposed to the Blakcbox Exporter charm.
+
+        Args:
+            charm: a `CharmBase` object which manages the
+                `BlackboxProbesProvider` object. Generally, this is `self` in
+                the instantiating class.
+            probes: the probes to configure in Blackbox Exporter passed as a list
+                    of configuration probes in python structures.
+            modules: an optional definition of modules for Blackbox Exporter to
+                use. For details on how to write pass a module, see the
+                [docs upstream](https://github.com/prometheus/blackbox_exporter/blob/master/CONFIGURATION.md).
+                Further examples are provided [upstream](https://github.com/prometheus/blackbox_exporter/blob/master/example.yml).
+            refresh_event: additional `CharmEvents` event (or a list of them)
+                on which the probes and modules should be updated.
+            relation_name: name of the relation providing the Blackbox Probes
+                service. It's recommended to not change it, to ensure a
+                consistent experience across all charms that use the library.
+        """
+        self.topology = JujuTopology.from_charm(charm)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._probes = [] if probes is None else probes
+        self._modules = {} if modules is None else modules
+
+        events = self._charm.on[self._relation_name]
+        self.framework.observe(events.relation_changed, self._set_probes_spec)
+
+        refresh_event = []
+        if not refresh_event:
+            if len(self._charm.meta.containers) == 1:
+                container = list(self._charm.meta.containers.values())[0]
+                refresh_event = [self._charm.on[container.name.replace("-", "_")].pebble_ready]
+        elif not isinstance(refresh_event, list):
+            refresh_event = [refresh_event]
+
+        module_name_prefix = f"juju_{self.topology.identifier}_"
+        self._prefix_probes(module_name_prefix)
+        self._prefix_modules(module_name_prefix)
+
+        self.framework.observe(events.relation_joined, self._set_probes_spec)
+        for ev in refresh_event:
+            self.framework.observe(ev, self._set_probes_spec)
+
+    def _set_probes_spec(self, _=None):
+        """Ensure probes target information is made available to Blackbox Exporter.
+
+        When a probes provider charm is related to a blackbox exporter charm, the
+        probes provider sets specification and metadata related to it
+        """
+        if not self._charm.unit.is_leader():
+            return
```

Comment:
I think the probes should be unit independent, all probes and module are set in the app relation because I expect to probe a list of URL or addresses that represent the service as a whole. Adding a comment to summarise this

Reply:
Added a comment on the docstring and added  by default the `refresh_event.append(self._charm.on.leader_elected)`, to list of events that trigger this function so that a new leader unit will get the list of probes on election.